libs/core/langchain_core/indexes/base.py

```diff
@@ -0,0 +1,369 @@
+    """Interface for a document index.
+
+    Example:
+        .. code-block:: python
+
+            from typing import Any, Iterator, List, Optional, Sequence, Tuple, Union, Iterable
+            from uuid import uuid4
+
+            from langchain_core.documents import Document
+            from langchain_core.indexes import UpsertResponse, DeleteResponse, Index
+
+            def uuid4_generator() -> Iterable[str]:
+                while True:
+                    yield str(uuid4())
+
+            class DictIndex(Index):
+
+                def __init__(self) -> None:
+                    self.store = {}
+
+                def upsert(
+                    self,
+                    documents: Iterable[Document],
+                    *,
+                    ids: Optional[Iterable[str]] = None,
+                    **kwargs: Any,
+                ) -> UpsertResponse:
+                    ids = ids or uuid4_generator()
+                    succeeded = []
+                    for id_, doc in zip(ids, documents):
+                        self.store[id_] = doc
+                        succeeded.append(id_)
+                    return UpsertResponse(succeeded=succeeded, failed=[])
+
+                def delete_by_ids(self, ids: Iterable[str]) -> DeleteResponse:
+                    succeeded = []
+                    failed = []
+                    for id_ in ids:
+                        try:
+                            del self.store[id_]
+                        except Exception:
+                            failed.append(id_)
+                        else:
+                            succeeded.append(id_)
+                    return DeleteResponse(succeeded=succeeded, failed=failed)
+
+                def lazy_get_by_ids(self, ids: Iterable[str]) -> Iterable[Document]:
+                    for id in ids:
+                        yield self.store[id]
+
+                def yield_keys(
+                    self, *, prefix: Optional[str] = None
+                ) -> Union[Iterator[str]]:
+                    prefix = prefix or ""
+                    for key in self.store:
+                        if key.startswith(prefix):
+                            yield key
+    """  # noqa: E501
+
+    @abstractmethod
+    def upsert(
+        self,
+        # TODO: Iterable or Iterator?
+        documents: Iterable[Document],
+        *,
+        ids: Optional[Iterable[str]] = None,
+        **kwargs: Any,
+    ) -> UpsertResponse:
+        """Upsert documents to index."""
+
+    # FOR CONTRIBUTORS: Overwrite this in Index child classes that support native async upsertions.
+    async def aupsert(
+        self,
+        documents: AsyncIterable[Document],
+        *,
+        ids: Optional[AsyncIterable[str]] = None,
+        **kwargs: Any,
+    ) -> UpsertResponse:
+        """Upsert documents to index. Default implementation, runs sync upsert() in async executor."""
+        # TODO: how to convert AsyncIterable -> Iterable
```

Comment:
Such a good question for a coding test interview 😄 !
I would say:
* with materialization (simple but not the most efficient): `[item async for item in async_iterable]`
* without materialization: I think a queue is needed. We can't use an async queue because it must be accessed from different threads. So using a sync queue in non-blocking mode, something like:
```python
async def async_it():
    for i in range(10):
        yield i

async def test():
    q = queue.Queue(2)
    sentinel = object()

    async def put_into_queue(item):
        while True:
            try:
                q.put(item, block=False)
                break
            except queue.Full:
                await asyncio.sleep(0)

    async def put_all_into_queue():
        async for item in async_it():
            await put_into_queue(item)

        await put_into_queue(sentinel)

    def sync_it():
        while True:
            try:
                item = q.get(block=False)
                if item is sentinel:
                    break
                yield item
            except queue.Empty:
                time.sleep(0)

    coro = run_in_executor(None, sum, sync_it())

    result = await asyncio.gather(put_all_into_queue(), coro)
    print(result[1])
```